src/app/modules/sgid/sgid.service.ts

```diff
@@ -0,0 +1,194 @@
+import { SgidClient } from '@opengovsg/sgid-client'
+import fs from 'fs'
+import { err, ok, Result, ResultAsync } from 'neverthrow'
+
+import { sgid } from '../../config/config'
+import { createLoggerWithLabel } from '../../config/logger'
+import { ApplicationError } from '../core/core.errors'
+
+import {
+  SgidCreateRedirectUrlError,
+  SgidFetchAccessTokenError,
+  SgidFetchUserInfoError,
+  SgidInvalidJwtError,
+  SgidInvalidStateError,
+  SgidVerifyJwtError,
+} from './sgid.errors'
+import { isSgidJwtPayload } from './sgid.util'
+
+const logger = createLoggerWithLabel(module)
+
+export class SgidService {
+  private client: SgidClient
+
+  private publicKey: string | Buffer
+
+  private cookieDomain: string
+  private cookieMaxAge: number
+  private cookieMaxAgePreserved: number
+
+  constructor({
+    cookieDomain,
+    cookieMaxAge,
+    cookieMaxAgePreserved,
+    privateKey,
+    publicKey,
+    ...sgidOptions
+  }: {
+    endpoint: string
+    clientId: string
+    clientSecret: string
+    privateKey: string
+    publicKey: string
+    redirectUri: string
+    cookieDomain: string
+    cookieMaxAge: number
+    cookieMaxAgePreserved: number
+  }) {
+    this.client = new SgidClient({
+      ...sgidOptions,
+      privateKey: fs.readFileSync(privateKey),
+    })
+    this.publicKey = fs.readFileSync(publicKey)
+    this.cookieDomain = cookieDomain
+    this.cookieMaxAge = cookieMaxAge
+    this.cookieMaxAgePreserved = cookieMaxAgePreserved
+  }
+
+  createRedirectUrl(state: string): Result<string, SgidCreateRedirectUrlError> {
```

Comment:
can we make it more explicit on what `state` is? From the call in the controller it seems to be made up of `formId` and `isPersistentLogin`; should we make those arguments of this function and create the state string in this function instead?

There will also be a benefit where the controller does not need to know how to create a state and leave that to this function.

Reply:
can we also have JSDocs for all the exported functions in this file